### PR TITLE
Websockets based implementation

### PR DESCRIPTION
**Pull request by omangbaheti**

### CODE REVIEW

**Comment by ahmed-shariff** [comment]
@omangbaheti look through the changes I made and see if it add up

File: Assets/Scripts/ViconNexusUnityStream/DataStreamer.cs [hunk 1]
```diff
@@ -1,56 +1,78 @@
...
     private Dictionary<string, Data> data = new();
     private Dictionary<string, string> rawData = new();
-    private async void Start()
+
+    /// <inheritdoc />
+    private void OnEnable()
+    {
+        SetupConnection();
+    }
+
```
Comment (omangbaheti) [inline Assets/Scripts/ViconNexusUnityStream/DataStreamer.cs]: Maybe move this into Awake() or Start() so this function is called only after the Custom Subject Scripts register their respective subjects.

File: Assets/Scripts/ViconNexusUnityStream/DataStreamer.cs [hunk 2]
```diff
@@ -60,14 +82,24 @@ private void StreamData(byte[] receivedData)
...
+    /// <summary>
+    /// Regsiter a subject to recieve subject data.
+    /// </summary>
+    public void RegisterSubject(string subjectName)
     {
-        webSocket.DispatchLatestMessage();
+        subjectList.Add(subjectName);
+        if (webSocket == null || webSocket.State == WebSocketState.Closed || webSocket.State == WebSocketState.Closing)
+        {
+            SetupConnection();
```
Comment (omangbaheti) [inline Assets/Scripts/ViconNexusUnityStream/DataStreamer.cs]: (logically it makes sense but i have a bad feeling about calling SetupConnection() here but its aight)

File: Assets/Scripts/ViconNexusUnityStream/SubjectDataManager.cs [hunk 3]
```diff
@@ -6,12 +6,39 @@
 using Newtonsoft.Json;
 using Newtonsoft.Json.Linq;
 
-public class DataStreamer : Singleton<DataStreamer>
+public class SubjectDataManager : MonoBehaviour
```
Comment (omangbaheti) [inline Assets/Scripts/ViconNexusUnityStream/SubjectDataManager.cs:9]: I am deeply saddened to learn about the loss of my dear friend, Mr. Singleton, who has been an invaluable support throughout my career as a developer. May Singleton rest in peace and find its new home in the hands of a naive developer, continuing to guide and inspire as it did for me.

File: Assets/Scripts/ViconNexusUnityStream/CustomSubjectScript.cs [hunk 4]
```diff
@@ -127,19 +119,19 @@ protected void SetupMessagePack()
...
         {
-            if (useDefaultData)
+            if (subjectDataManager.UseDefaultData)
             {
                 ProcessData(defaultDataObj, defaultData);
             }
             else
             {
-                ProcessData(dataStreamer.StreamedData[subjectName], dataStreamer.StreamedRawData[subjectName]);
+                ProcessData(subjectDataManager.StreamedData[subjectName], subjectDataManager.StreamedRawData[subjectName]);
```
Comment (omangbaheti) [inline Assets/Scripts/ViconNexusUnityStream/CustomSubjectScript.cs]: Maybe this if-else logic should be managed internally within subjectDataManager. CustomSubject should remain unaware of the default data status.

File: Assets/Scripts/ViconNexusUnityStream/DataStreamer.cs (hunk 1 above)
Comment (ahmed-shariff) [reply Assets/Scripts/ViconNexusUnityStream/DataStreamer.cs]: That's a good point. OnEnable should check the registered subjects before setting things up.

File: Assets/Scripts/ViconNexusUnityStream/DataStreamer.cs (hunk 2 above)
Comment (ahmed-shariff) [reply Assets/Scripts/ViconNexusUnityStream/DataStreamer.cs]: why?

File: Assets/Scripts/ViconNexusUnityStream/SubjectDataManager.cs (hunk 3 above)
Comment (ahmed-shariff) [reply Assets/Scripts/ViconNexusUnityStream/SubjectDataManager.cs:9]: RIP

File: Assets/Scripts/ViconNexusUnityStream/CustomSubjectScript.cs (hunk 4 above)
Comment (ahmed-shariff) [reply Assets/Scripts/ViconNexusUnityStream/CustomSubjectScript.cs]: That also requires rewriting how the data gets written out (another TODO). We'll circle back to that when we know this works.

File: Assets/Scripts/ViconNexusUnityStream/DataStreamer.cs (hunk 2 above)
Comment (omangbaheti) [reply Assets/Scripts/ViconNexusUnityStream/DataStreamer.cs]: tbh its an irrational ick

but i feel like RegisterSubject() should not automatically call SetupConnection() and that it should be handled in awake or start

**Comment by ahmed-shariff** [comment]
Fixed a few more things in both repos. If all good, I can merge em @omangbaheti

**Comment by omangbaheti** [comment]
> Fixed a few more things in both repos. If all good, I can merge em @omangbaheti

Saul Goodman 👍

**Pull request merged**